src/compute/src/sink/persist_sink.rs

```diff
@@ -130,111 +593,471 @@ where
-                            .or_default()
-                            .push(((SourceData(Ok(row)), ()), ts, diff));
+                        let is_new = in_flight_descriptions.insert(batch_description.clone());
+
+                        assert!(
+                            is_new,
+                            "append_batches: sink {} got more than one batch \
+                            for a given description in-flight: {:?}",
+                            sink_id, in_flight_batches
+                        );
                     }
                 });
 
-                err_input.for_each(|_cap, data| {
-                    data.swap(&mut err_buffer);
+                // Ingest new written batches
+                batches_input.for_each(|_cap, data| {
+                    data.swap(&mut batch_buffer);
+                    for batch in batch_buffer.drain(..) {
+                        let batch = write.batch_from_hollow_batch(batch);
+                        let batch_description = (batch.lower().clone(), batch.upper().clone());
+
+                        let batches = in_flight_batches
+                            .entry(batch_description)
+                            .or_insert_with(Vec::new);
 
-                    for (error, ts, diff) in err_buffer.drain(..) {
-                        stash
-                            .entry(ts)
-                            .or_default()
-                            .push(((SourceData(Err(error)), ()), ts, diff));
+                        batches.push(batch);
                     }
                 });
 
-                let mut updates = stash
+                // Peel off any batches that are not beyond the frontier
+                // anymore.
+                //
+                // It is correct to consider batches that are not beyond the
+                // `batches_frontier` because it is held back by the writer
+                // operator as long as a) the `batch_description_frontier` did
+                // not advance and b) as long as the `desired_frontier` has not
+                // advanced to the `upper` of a given batch description.
+
+                let mut done_batches = in_flight_descriptions
                     .iter()
-                    .filter(|(ts, _updates)| !input_frontier.less_equal(ts))
-                    .flat_map(|(_ts, updates)| updates.iter());
-
-                if PartialOrder::less_than(&*shared_frontier.borrow(), &input_frontier) {
-                    // We always append, even in case we don't have any updates, because appending
-                    // also advances the frontier.
-                    let lower = shared_frontier.borrow().clone();
-                    // TODO(aljoscha): Figure out how errors from this should be reported.
-                    write
-                        .append(updates, lower, input_frontier.clone())
+                    .filter(|(lower, _upper)| !PartialOrder::less_equal(batches_frontier, lower))
```

Comment:
Out of curiosity: How do you decide when to use the negated version over the equivalent non-negated one (`PartialOrder::less_than(lower, batches_frontier)` in this case). Assuming I'm not missing something and they are indeed equivalent, I would always consider the non-negated one more readable. But that might be because my brain has not yet adapted to timely :)

Reply:
Here sometimes my old ways with `Antichain::less_than` come through. With `Antichain` and a single timestamp you can't just do the inverse comparison. So I don't have a good answer here and I might turn it into `PartialOrder::less_than(lower, batches_frontier)`